app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,146 @@
+     */
+    public void removeDevice(String deviceId) {
+        if (deviceFolderMap.containsKey(deviceId)) {
+            deviceFolderMap.remove(deviceId);
+        }
+    }
+
+    /**
+     * Adds a folder to the cache model if it does not exist.
+     */
+    private void addFolder(String deviceId, String folderId) {
+        if (!deviceFolderMap.containsKey(deviceId)) {
+            addDevice(deviceId);
+        }
+
+        if (!deviceFolderMap.get(deviceId).containsKey(folderId)) {
+            deviceFolderMap.get(deviceId).put(folderId, new CompletionInfo());
+        }
+    }
+
+    /**
+     * Removes a folder from the cache model.
+     */
+    public void removeFolder(String folderId) {
+        for (String deviceId : deviceFolderMap.keySet()) {
+            if (deviceFolderMap.get(deviceId).containsKey(folderId)) {
+                deviceFolderMap.get(deviceId).remove(folderId);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Updates device information in the cache model.
+     */
+    public void updateDevices(List<Device> newDeviceSet) {
+        List<String> removedDevices = new ArrayList<>();;
+        Boolean deviceFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            deviceFound = false;
+            for (Device device : newDeviceSet) {
+                if (device.deviceID.equals(deviceId)) {
+                    deviceFound = true;
+                    break;
+                }
+            }
+            if (!deviceFound) {
+                removedDevices.add(deviceId);
+            }
+        }
+        for (String deviceId : removedDevices) {
+            Log.v(TAG, "updateDevices: Remove device '" + deviceId + "' from cache model");
+            removeDevice(deviceId);
+        }
+    }
+
+    /**
+     * Updates folder information in the cache model.
+     */
+    public void updateFolders(List<Folder> newFolderSet) {
+        List<String> removedFolders = new ArrayList<>();;
+        Boolean folderFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            for (String folderId : deviceFolderMap.get(deviceId).keySet()) {
+                folderFound = false;
+                for (Folder folder : newFolderSet) {
+                    if (folder.id.equals(folderId)) {
+                        folderFound = true;
+                        break;
+                    }
+                }
+                if (!folderFound) {
+                    removedFolders.add(folderId);
+                }
+            }
+        }
+        for (String folderId : removedFolders) {
+            Log.v(TAG, "updateFolders: Remove folder '" + folderId + "' from cache model");
+            removeFolder(folderId);
+        }
+    }
+
+    /**
+     * Calculates remote device sync completion percentage across all folders
+     * shared with the device.
+     */
+    public int getDeviceCompletion(String deviceId) {
+        addDevice(deviceId);
+        int folderCount = 0;

```

Comment:
Instead of calling add and what not, you could do getOrDefault and just potentoally return zeroes early.

Reply:
I tried implementing this, but gradle complains about the API level must be raised to >= 24 to use getOrDefault.  ../../src/main/java/com/nutomic/syncthingandroid/model/Completion.java:122: Call requires API level 24 (current min is 16): java.util.HashMap#getOrDefault

Reply:
Trying implementation via extending HashMap to DefaultHashMap ...

Reply:
... done.